### PR TITLE
Narrow deps to snap-core, plus readme for snaplets

### PR DESCRIPTION
**Pull request by sirlensalot**
Hi! This drops the full 'snap' dependency in favor of a more minimal 'snap-core' dep, the snap framework brings in a huge amount of bloat for those of us doing simpler servers that nonetheless need CORS support. Given that snaplet support comes down to issuing "wrapSite" it seems like an unobtrusive change, and I've made a README to document this. Thanks! Stuart

### CODE REVIEW

**Comment by ocharles** [comment]
Hi. This came up before - https://github.com/ocharles/snap-cors/issues/6 - and I'm still not sure I really want to make such a breaking change now. Calling it unobtrusive isn't true - it's a major API change that would impact downstream users. Exactly how many use it with `snap` rather than `snap-core` I unfortunately don't know.

What particular problem is the large amount of dependencies causing you?

**Comment by slpopejoy** [comment]
Dep pain is felt in sandbox setup time and CI mainly. Building snap-cors with full snap takes 3m4s longer than with just snap-core, on my recent MB pro. That's painful.

Agree it is breaking, but it's so easy to fix. Fixing TH code for GHC 7.10 was WAY worse :). I think breaking changes are OK if they're justified and easily resolvable.

My bigger beef is snap-cors forces me to abandon my choice to depend on just snap-core and snap-server. The 'snap' package buys you into a whole framework, with heist, snaplets etc... All that just to get one line of snaplet integration for CORS.

**Comment by ocharles** [comment]
Right, but those deps only need to be installed once and can be cached
heavily if you use something like Stack on Nix. I wouldn't imagine that you
should have to really install `snap` more than once these days. I get your
frustration, but me changing my package will not solve your problems -
having a wide spanning set of dependencies is just what happens in
day-to-day Haskell programming these days, and there are other solutions.
Even if I were to make this change, it won't be practical for you to use
for quite a while, as I would like to go through a deprecation period first
(informing downstream users how to prepare their code), so you'd only see
these changes in a month or so...

I think unfortunately your best option might be to just fork this Github
repo and have your own version without the Snap dependency. As this project
is pretty much finished, I doubt you'll have many merge troubles as time
goes on.

**Comment by slpopejoy** [comment]
> me changing my package will not solve your problems

That's not true -- I have multiple production projects depending on snap-cors who would enjoy faster builds and less deps with this change. I don't use Stack but even if I did there are reasons for clean-env builds, e.g., container (docker) builds, that would nonetheless be slower with more deps.

Deprecation is a responsible approach, and a month-or-so timeframe is fine ... but perhaps overly cautious. FWIW, a google search for `applyCORS` gets hits in haskell code while `wrapCORS` doesn't get anything. My anecdotal view is snap-core/server is more popular than full snap, with people reaching for yesod when they want batteries-included.

In summary, I think the dep-shave is a valid goal due to snap-core/server being widely-used  independent of snap.

**Comment by gregorycollins** [comment]
We can fold CORS support into the Snap.Util.* namespace in snap-core for the upcoming snap version 1.0 if someone sends the right PR. I don't have time to work on it beyond that.

I agree with OP that a snap-core dependency would be preferable here.

**Comment by ocharles** [comment]
I've come around on this. No one has spoken about actually needing `snap` support (other than me, and I'm not even using this atm). The real long term path is probably to get this into `snap-core`, as @gregorycollins mentions.

**Comment by slpopejoy** [comment]
Regarding merging into Snap, I made a `snap-server` patch for this -- https://github.com/snapframework/snap-server/pull/97 . It adds deps, so it seemed like `snap-server` was the better place for it. @gregorycollins if you think this should go into core let me know.

**Comment by slpopejoy** [comment]
`snap-core` migration is done! https://hackage.haskell.org/package/snap-core-1.0.2.0/docs/Snap-Util-CORS.html
